### PR TITLE
[browser][wasm] simplify runtime bindings and improve error messages

### PR DESCRIPTION
**Pull request by lewing**
Draft to do Runtime binding by name and add more descriptive errors.

### CODE REVIEW

File: src/mono/wasm/runtime/binding_support.js [hunk 1]
```diff
@@ -88,40 +88,54 @@ var BindingSupportLib = {
...
+			bind_method ("GetDotNetObject", "im");
+
+			bind_method ("BoxInt", "im");
+			bind_method ("BoxDouble", "dm");
+			bind_method ("BoxBool", "im");
+			bind_method ("IsSimpleArray", "mi");
+			bind_method ("SetupJSContinuation", "mo")
+
+			bind_method ("CreateTaskSource", "i");
+			bind_method ("SetTaskSourceFailure", "oo");
```
Comment (radical) [inline src/mono/wasm/runtime/binding_support.js:125]: `s,SetTaskSourceFailure,SetTaskSourceResult,`

**Review by radical — CHANGES_REQUESTED** [review]

File: src/mono/wasm/runtime/binding_support.js (hunk 1 above)
Comment (lewing) [reply src/mono/wasm/runtime/binding_support.js:125]: good catch, clearly we don't have tests for this

File: src/mono/wasm/runtime/binding_support.js [hunk 2]
```diff
@@ -88,40 +88,54 @@ var BindingSupportLib = {
...
-				var res = BINDING.find_method (wasm_runtime_class, method_name, -1)
-				if (!res)
-					throw "Can't find method " + namespace + "." + classname + ":" + method_name;
-				return res;
+			this.runtime = {
+				class: wasm_runtime_class
+			};
+
+			var bind_method = function(method_name, signature, namespace = BINDING.runtime) {
+				var method = BINDING.find_method (namespace.class, method_name, -1);
```
[the remark this answers is not in the document]
Comment (lewing) [reply src/mono/wasm/runtime/binding_support.js:96]: still have some changes pending here

**Comment by srxqds** [comment]
why this pr not push to mono repo?

**Comment by lewing** [comment]
@srxqds because this part of the code doesn't automatically sync and while we might manually backport it, it is still a work in progress

File: src/mono/wasm/runtime/binding_support.js [hunk 3]
```diff
@@ -66,62 +65,60 @@ var BindingSupportLib = {
 			// receives a byteoffset into allocated Heap with a size.
 			this.mono_typed_array_new = Module.cwrap ('mono_wasm_typed_array_new', 'number', ['number','number','number','number']);
 
-			var binding_fqn_asm = this.BINDING_ASM.substring(this.BINDING_ASM.indexOf ("[") + 1, this.BINDING_ASM.indexOf ("]")).trim();
-			var binding_fqn_class = this.BINDING_ASM.substring (this.BINDING_ASM.indexOf ("]") + 1).trim();
-			
-			this.binding_module = this.assembly_load (binding_fqn_asm);
+			var binding_info = this.parse_fqn (this.BINDING_ASM);
```
Comment (radical) [inline src/mono/wasm/runtime/binding_support.js:68]:
```suggestion
			const binding_info = this.parse_fqn (this.BINDING_ASM);
```

File: src/mono/wasm/runtime/binding_support.js [hunk 4]
```diff
@@ -66,62 +65,60 @@ var BindingSupportLib = {
...
-			{
-				var namespace = "System.Runtime.InteropServices.JavaScript";
-				var classname = binding_fqn_class.length > 0 ? binding_fqn_class : "Runtime";
-				if (binding_fqn_class.indexOf(".") != -1) {
-					var idx = binding_fqn_class.lastIndexOf(".");
-					namespace = binding_fqn_class.substring (0, idx);
-					classname = binding_fqn_class.substring (idx + 1);
-				}
-			}
+			var runtime_class = this.find_class (this.binding_module, binding_info.namespace, binding_info.class)
```
Comment (radical) [inline src/mono/wasm/runtime/binding_support.js:74]:
```suggestion
			const runtime_class = this.find_class (this.binding_module, binding_info.namespace, binding_info.class)
```

File: src/mono/wasm/runtime/binding_support.js [hunk 5]
```diff
@@ -740,15 +703,17 @@ var BindingSupportLib = {
...
-			
+
+		parse_fqn: function (fqn) {
 			var assembly = fqn.substring(fqn.indexOf ("[") + 1, fqn.indexOf ("]")).trim();
 			fqn = fqn.substring (fqn.indexOf ("]") + 1).trim();
 
-			var methodname = fqn.substring(fqn.indexOf (":") + 1);
-			fqn = fqn.substring (0, fqn.indexOf (":")).trim ();
+			var methodstart = fqn.indexOf (":");
+			var methodname = "";
```
Comment (radical) [inline src/mono/wasm/runtime/binding_support.js:712]:
```suggestion
			const methodstart = fqn.indexOf (":");
			let methodname = "";
```

File: src/mono/wasm/runtime/binding_support.js [hunk 6]
```diff
@@ -740,15 +703,17 @@ var BindingSupportLib = {
...
+			if (methodstart != -1) {
+				methodname = fqn.substring(fqn.indexOf (":") + 1);
+				fqn = fqn.substring (0, fqn.indexOf (":")).trim ();
```
Comment (radical) [inline src/mono/wasm/runtime/binding_support.js:715]:
```suggestion
				methodname = fqn.substring(methodstart + 1);
				fqn = fqn.substring (0, methodstart).trim ();
```

File: src/mono/wasm/runtime/binding_support.js [hunk 7]
```diff
@@ -758,17 +723,30 @@ var BindingSupportLib = {
...
+				class: classname,
+				method: methodname
+			};
+		},
+
+		resolve_method_fqn: function (fqn) {
+			this.bindings_lazy_init ();
+
+			var info = this.parse_fqn (fqn);
+			var asm = this.assembly_load (info.assembly);
```
Comment (radical) [inline src/mono/wasm/runtime/binding_support.js:738]:
```suggestion
			const info = this.parse_fqn (fqn);
			const asm = this.assembly_load (info.assembly);
```

File: src/mono/wasm/runtime/binding_support.js [hunk 8]
```diff
@@ -758,17 +723,30 @@ var BindingSupportLib = {
...
+				throw new Error (`Could not find assembly: ${info.assembly}`);
 
-			var klass = this.find_class(asm, namespace, classname);
+			var klass = this.find_class(asm, info.namespace, info.class);
 			if (!klass)
-				throw new Error ("Could not find class: " + namespace + ":" +classname);
+				throw new Error (`Could not find class: ${info.namespace}:${info.class}`);
 
-			var method = this.find_method (klass, methodname, -1);
+			var method = this.find_method (klass, info.method, -1);
```
Comment (radical) [inline src/mono/wasm/runtime/binding_support.js:746]:
```suggestion
			const method = this.find_method (klass, info.method, -1);
```

**Review by radical — CHANGES_REQUESTED** [review]

**Pull request closed without merging**